externs/browser/window.js

```diff
@@ -181,22 +181,32 @@ function prompt(message, opt_value) {}
 function setImmediate(callback, callbackParams) {}
 
 /**
- * @param {Function|!TrustedScript|string} callback
+ * @typedef {function(...*):undefined}
```

Comment:
An even more basic issue is that tightening up callback type is catching lots of new errors, which aren't immediately clear if they are valid or not.  For example, in closure-library, we have errors like:

closure/goog/module/modulemanager.js:646: ERROR - inconsistent return type
found   : Error
required: undefined
          () => new Error('Module already loaded: ' + id), 0);
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

closure/goog/module/modulemanager.js:820: ERROR - actual parameter 1 of Window.setTimeout does not match formal parameter
found   : function(*): undefined
required: (TrustedScript|function(...*): undefined|null|string)
      window.setTimeout(goog.bind(callbackWrapper.execute, callbackWrapper), 0);
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

I somewhat suspect that the new type is to restrictive.  Can you investigate?

Reply:
I will investigate over the weekend

Reply:
The first issue highlighted seems to be a bug in `modulemanager.js`. Looking at the [goog.module.ModuleManager.prototype.processModulesForLoad_](https://github.com/google/closure-library/blob/0f3a8a51bb4897086642bf9934270cd3d6a23ff2/closure/goog/module/modulemanager.js#L638-L647) I believe that they are trying to throw an error on the next task. Similar code exists at [goog.async.throwException](https://github.com/google/closure-library/blob/2159b6717c6b6869a780eafc6ae0fb0737599f40/closure/goog/async/nexttick.js#L43-L46). In this scenario I believe the type checker has identified a bug as the code as it stands is effectively a noop.

Looking into the second scenario now ...

Reply:
The second scenario is because I did not understand the typing system correctly it seems. I had assumed that `function(...*):undefined` would accept a function typed with `function():undefined` but that seems not to be the case. I ended up having to add both `function(...*):undefined` and `function():undefined` into type union. (Although it is possible that I am doing something else wrong here...).

Anyhoo the test code I experiemented with was

**foo.extern.js:**
```javascript
/**
 * @fileoverview BLAH
 * @externs
 */

/**
 * @typedef {function(...*):undefined}
 */
var MyTimerCallbackFunction;

/**
 * @typedef {function():undefined}
 */
var MySimpleCallbackFunction;

/**
 * @typedef {!MyTimerCallbackFunction|!TrustedScript|string|!MySimpleCallbackFunction}
 */
var MyTimerCallback;

/**
 * @param {?MyTimerCallback} callback
 * @param {number=} opt_delay
 * @param {...*} callbackParams
 * @return {number}
 */
function mySetTimeout(callback, opt_delay, callbackParams) {}

```
**foo.js:**
```javascript
/** @type {function(): undefined} */
const myCallback = () => console.log('hi');

/** @type {function(?): undefined} */
const myCallback2 = (count) => console.log(count);

const foo = {};

const myOtherCallback = myCallback2.bind(foo, 3)

mySetTimeout(myCallback, 0);
mySetTimeout(myCallback2, 0, 33);
mySetTimeout(myOtherCallback, 0);
```

If you remove `MySimpleCallbackFunction` from the extern then `mySetTimeout(myCallback2, 0, 33)` will not type check with.

```
foo.js:12: WARNING - actual parameter 1 of mySetTimeout does not match formal parameter
found   : function(?): undefined
required: (TrustedScript|function(...*): undefined|null|string)
mySetTimeout(myCallback2, 0, 33);
```

It is not entirely clear why it behaves this way ...

Reply:
I'm assuming that the requirement has to do with contravariance and not wanting to allow a function that can only accept one argument to substitute for one that can handle arbitrarily many arguments.  (See http://closuretools.blogspot.com/2012/06/subtyping-functions-without-poking-your.html for some good background)

So I think that adding MySimpleCallbackFunction only hides the issue for the 1-argument case, but not for all the other cases (1-argument, 2-argument, etc).

So I don't see a feasible annotation to put here that is better than !Function, unfortunately.

Reply:
Right. I will rollback to `!Function` and leave this as something that is patched downstream in Elemental2.

Thanks for looking into it.